source/code/go/src/plugins/telemetry.go

```diff
@@ -0,0 +1,155 @@
+)
+
+var (
+	// FlushedRecordsCount indicates the number of flushed records in the current period
+	FlushedRecordsCount float64
+	// FlushedRecordsTimeTaken indicates the cumulative time taken to flush the records for the current period
+	FlushedRecordsTimeTaken float64
+	// CommonProperties indicates the dimensions that are sent with every event/metric
+	CommonProperties map[string]string
+	// TelemetryClient is the client used to send the telemetry
+	TelemetryClient appinsights.TelemetryClient
+	// ContainerLogTelemetryTicker sends telemetry periodically
+	ContainerLogTelemetryTicker *time.Ticker
+)
+
+const (
+	clusterTypeACS               = "ACS"
+	clusterTypeAKS               = "AKS"
+	controllerTypeDaemonSet      = "DaemonSet"
+	controllerTypeReplicaSet     = "ReplicaSet"
+	envAKSResourceID             = "AKS_RESOURCE_ID"
+	envACSResourceName           = "ACS_RESOURCE_NAME"
+	envAppInsightsAuth           = "APPLICATIONINSIGHTS_AUTH"
+	metricNameAvgFlushRate       = "ContainerLogAvgRecordsFlushedPerSec"
+	defaultTelemetryPushInterval = 300
+
+	// EventNameContainerLogInit name of the event
+	EventNameContainerLogInit = "ContainerLogPluginInitialized"
+)
+
+// Initialize initializes the telemetry artifacts
+func initialize(telemetryPushIntervalProperty string, agentVersion string) (int, error) {
+
+	telemetryPushInterval, err := strconv.Atoi(telemetryPushIntervalProperty)
+	if err != nil {
+		Log("Error Converting telemetryPushIntervalProperty %s. Using Default Interval... %d \n", telemetryPushIntervalProperty, defaultTelemetryPushInterval)
+		telemetryPushInterval = defaultTelemetryPushInterval
+	}
+
+	ContainerLogTelemetryTicker = time.NewTicker(time.Second * time.Duration(telemetryPushInterval))
+
+	encodedIkey := os.Getenv(envAppInsightsAuth)
+	if encodedIkey == "" {
+		Log("Environment Variable Missing \n")
+		return -1, errors.New("Missing Environment Variable")
+	}
+
+	decIkey, err := base64.StdEncoding.DecodeString(encodedIkey)
+	if err != nil {
+		Log("Decoding Error %s", err.Error())
+		return -1, err
+	}
+
+	TelemetryClient = appinsights.NewTelemetryClient(string(decIkey))
+
+	CommonProperties = make(map[string]string)
+	CommonProperties["Computer"] = Computer
+	CommonProperties["WorkspaceID"] = WorkspaceID
+	CommonProperties["ControllerType"] = controllerTypeDaemonSet
+	CommonProperties["AgentVersion"] = agentVersion
+
+	aksResourceID := os.Getenv(envAKSResourceID)
+	// if the aks resource id is not defined, it is most likely an ACS Cluster
+	if aksResourceID == "" {
+		CommonProperties["ACSResourceName"] = os.Getenv(envACSResourceName)
+		CommonProperties["ClusterType"] = clusterTypeACS
+
+		CommonProperties["SubscriptionID"] = ""
+		CommonProperties["ResourceGroupName"] = ""
+		CommonProperties["ClusterName"] = ""
+		CommonProperties["Region"] = ""
+
+	} else {
+		CommonProperties["ACSResourceName"] = ""
+		splitStrings := strings.Split(aksResourceID, "/")
+		CommonProperties["SubscriptionID"] = splitStrings[2]
```

Comment:
this is dangerous. assuming array will have all. Can we check bounds before we do this assignment ?